Add basic usage docs for feature flags (#22069)
sourcegraph/sourcegraph: doc/dev/how-to/use_feature_flags.md

They are designed to be useful for A/B testing, and the values of all active feature flags
are added to every event log for the purpose of analytics.

## How it works

Each feature flag is either a static feature flag, or a "rollout" flag. 
- A static feature flag has a single value (currently only "true" or "false") for all users that haven't overriden it.
- A rollout flag assigns a random (but stable) value to each user. Each rollout flag is created with a percentage of users that should be randomly assigned the value "true".

A user is identified either by their user ID (if logged in), or by an anonymous user ID in local 
storage. 

The set of evaluated feature flags is appended to each event log so they can be queried against
for analytics.

## Lifecycle of a feature flag (for A/B testing)

The standard use of a feature flag for A/B testing will look like the following:

1) Implement the feature that you want to be behind a feature flag
2) Deploy to sourcegraph.com
3) Create the feature flag through the GraphQL API
4) Measure the effect of the feature flag
5) Disable or delete the feature flag
6) Remove the code that references the feature flag

> ### Implement the feature flag

Reviewer: We should update this section to match the work @novoselrok  did to make feature flags type safe. See https://github.com/sourcegraph/sourcegraph/pull/22121/files

Reviewer: @novoselrok  can you update this PR with the new type-safe map usage instead?

# How to use feature flags

> This document will take you through how to add, remove, and modify feature flags.

Reviewer: May be useful to define feature flags (in contrast with the existing "experimental features" that are already present as part of the settings).

## When to use feature flags

Feature flags, as opposed to experimental features, are intended to be strictly short-lived.
They are designed to be useful for A/B testing, and the values of all active feature flags
are added to every event log for the purpose of analytics.

## How it works

Each feature flag is either a static feature flag, or a "rollout" flag. 
- A static feature flag has a single value (currently only "true" or "false") for all users that haven't overriden it.
- A rollout flag assigns a random (but stable) value to each user. Each rollout flag is created with a percentage of users that should be randomly assigned the value "true".

A user is identified either by their user ID (if logged in), or by an anonymous user ID in local 
storage. 

The set of evaluated feature flags is appended to each event log so they can be queried against
for analytics.

## Lifecycle of a feature flag (for A/B testing)

The standard use of a feature flag for A/B testing will look like the following:

1) Implement the feature that you want to be behind a feature flag
2) Deploy to sourcegraph.com

> 3) Create the feature flag through the GraphQL API

Reviewer 2: Dos this mean all feature flag logic are assumed to live client side? If so, "when to use feature flags" might mention that feature flags are intended for frontend UI/UX testing.

Author: No -- this is intended to be a representative example of the flow, but feature flags are also available in the backend. I'll make that more clear and add an example of using it in the backend

4) Measure the effect of the feature flag
5) Disable or delete the feature flag
6) Remove the code that references the feature flag

### Implement the feature flag

In the frontend, evaluated feature flags for the current user are available on 
the SourcegraphWebAppState. These can be prop-drilled into the components that need access to them.

> Whenever a flag is used from a flags set, it should be provided a default so that 
> the code can be deployed before creating the feature flag, the feature flag can safely 
> be deleted before removing the referencing code, and so enterprise deployments work as
> expected. 

Reviewer 2: ```suggestion
Ensure that a default value is set for feature flags so that 
(i) code can be deployed before creating the feature flag, (ii) deleting the feature flag is safe before removing referenced code, (iii) enterprise deployments continue to work as
expected. 
```

Just some word smithing. Don't feel strongly.

For example:
```typescript
let myFeatureFlagValue = this.state.featureFlags.myFeatureFlag || false
```

### Create the feature flag through the GraphQL API

To create a rollout feature flag, currently the best way is to use the GraphQL API.

Go to `sourcegraph.com/api/console`, then create a feature flag like the following:
```graphql
mutation CreateFeatureFlag{
  createFeatureFlag(
    name: "myFeatureFlag",
    rolloutBasisPoints: 5000,
  ){
    __typename
  }
}
```

> The value of `rolloutBasisPoints` is measured in increments of 0.01% (a basis point), 
> for to create a feature flag that applies to 50% of users, set `rolloutBasisPoints` 
> to 5000.

Reviewer 2: ```suggestion
The value of `rolloutBasisPoints` is measured in increments of 0.01% (a basis point).
To create a feature flag that applies to 50% of users, set `rolloutBasisPoints` 
to 5000.
```

	JSON_VALUE(feature_flags, '$.myFeatureFlag') AS my_flag, 
	count(*) 
FROM `telligentsourcegraph.dotcom_events.events` 
WHERE name = 'ShareButtonClicked' 
GROUP BY my_flag;
```

### Disable or delete the feature flag

In most cases, after an A/B test is performed, a feature flag should be deleted.
That can be done with a GraphQL query like the following:
```graphql
mutation DeleteFeatureFlag{
  deleteFeatureFlag(
    name: "myFeatureFlag",
  ){
    __typename
  }
}
```

Once a feature flag is deleted, it will no longer be added to events as metadata,
so removing the code path that uses it will not change any measurements.

### Remove the code that references the feature flag

Once the feature flag is deleted, the code that references it can be safely deleted
without changing any of the measurements. 

## Feature flag overrides

In addition to feature flags as described above, you can also create feature flag
overrides. This is useful if you'd like to test a feature flag locally by assigning
your user a specific value, or if you'd like to do an A/B test on members of the 
Sourcegraph org. 

Overrides can either apply to a single user or an entire org. If both are set, a user

> override takes precedent over an org override.

Reviewer 2: ```suggestion
override takes precedence over an org override.
```

Author: 🤦
